api/src/main/java/org/apache/iceberg/SortField.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.Objects;
+import org.apache.iceberg.SortTransforms.SortTransform;
+
+public class SortField implements Serializable {
+  public enum Direction {
+    ASC, DESC
+  }
+
+  public enum NullOrder {
+    NULLS_FIRST, NULLS_LAST
+  }
+
+  private final String name;
+  private final Direction direction;
+  private final NullOrder nullOrder;
+  private final SortTransform transform;
```

Comment:
I think it is likely that sort transforms and partition transforms mostly overlap, so we may consider using the same `Transform` class for both.

When configuring a table's sort order, I think we want to allow both an order across tasks as well as a local order in a single task. Then, the table can be configured with an ordering that includes partition columns so that Iceberg can tell the processing engine to use an ordering for a write that includes partitions.

A concrete use case is that we have a table partitioned by `date(ts)` and `bucket(id, 16)`, and we want Spark to write using `date(ts)`, `bucket(id, 16)`, and finally, `id`. If we want partitions in the sort order, then we will need to support all of those transforms and probably want to standardize on `Transform`.

Reply:
I agree that sharing transforms between `SortOrder` and `PartitionSpec` is a good idea.

W.r.t the order within/across tasks. Did I get correctly the idea is that query engines should sort data globally/locally depending on whether partition columns are included in the sort order? If yes, I have a couple of questions. For example, we can have multiple pipelines writing to the same table. While it is beneficial to sort data globally in batch jobs, we might want to sort data within tasks or disable sorting at all in streaming jobs. Internally, we allow our customers to configure their sort mode on write via a SQL config. That gives them the needed flexibility to configure the sort mode per job. Since we plan to annotate files with `sort_order_id`, compaction jobs might take that into account as well. Is it reasonable to say the sort mode should be configurable per job rather than per table?

My original understanding was that `SortOrder` should apply only within partitions. Basically, the actual sort key is derived by combining partition and sort values (which should be done by query engines). Also, not including partition columns in the sort key won’t work because we will generate a lot of files per task. Is this reasonable?

Reply:
Iceberg should support configuring an order that gets applied to all jobs writing to a table. We want to enable building services that can make changes to a table, like the file format, compression, or sort order used. Embedding the sort order in a job requires having a human change the job and not just table configuration because it requires editing SQL or rewriting a DataFrame operation. Sort order should be declarative, not imperative.

You're right that there are times want to have control over a write in a job. For example, you may want to write unsorted data to meet an SLA, but then have a service go rewrite that data with a more optimal sort order. I think this need is less common and the default should be to use the sort order from the table's configuration. We just need a way to override the table configuration.

Also, table configuration should support both local and global ordering. Our best practice is to use a global ordering to handle skew, prepare data for filters, and make the data encode or compress better. But, maybe this should be configured separately, like Hive's `DISTRIBUTE BY` and `SORT BY`. What do you think?

Reply:
I think `DISTRIBUTE BY` and `SORT BY` can be potentially quite flexible. As those clauses have specific meanings in Hive/Spark, we probably need to support `ORDER BY` and `CLUSTER BY` (optionally) in order to express arbitrary sort orders. For example,  we cannot express a global sort order with `DISTRIBUTE BY` and `SORT BY`.

We can have something like below:

**Case 1**

Iceberg:
```
SortOrder.builderFor(schema)
    .distributeBy("part_col")
    .distributeBy(bucket("ts", 16))
    .sortBy("id")
    .build()
```
Query engines:

```
Sort("id", global = false)
    HashPartitioning(bucket("ts", 16), "part_col")
```

**Case 2**

Iceberg:
```
SortOrder.builderFor(schema)
    .orderBy("part_col_1")
    .orderBy("part_col_2")
    .orderBy("id")
    .build()
```

Query engines:
```
Sort("part_col_1", "part_col_2", "id", global = true)
```

**Case 3**

Iceberg:
```
SortOrder.builderFor(schema)
    .clusterBy("part_col_1")
    .clusterBy("part_col_2")
    .sortBy("id")
    .build()
```

Query engines:
```
Sort("part_col_1", "part_col_2", "id", global = false)
    HashPartitioning("part_col_1", "part_col_2")
```

Reply:
In Spark, we will need a way how to define a sort order in DDLs as well as to extend DS V2 so that tables can report their ordering. @rdblue, what do you think?

Reply:
In Spark's DDL, `CLUSTERED BY` basically adds bucketing to a table, or a partition that uses a bucket function in v2. `SORTED BY` is only allowed in DDL with a `CLUSTERED BY` clause. It looks like you're using `clusteredBy` a little differently here, to express a way to distribute and cluster.

We've also discussed this a bit on the Spark dev list. I filed [SPARK-23889](https://jira.apache.org/jira/browse/SPARK-23889) with the proposal from that discussion. The proposal was to add a required ordering that is global unless there is also a required distribution (called "clustering", but let's use "distribution" because of the confusion with `CLUSTERED BY`). Basically, the ordering produces a default distribution, but that can be overridden; when overridden the ordering is local.

I think that same approach would work here. I like the idea of using "distributed by" to express the required distribution, and I think it would work well to make the ordering determine the distribution unless it is overridden.

That approach fits nicely with standard SQL, which doesn't have separate `SORT BY` and `ORDER BY` that mean different things (which I think would be confusing). By default, adding an `ORDER BY` means a global ordering, but you can also add a `DISTRIBUTE BY` to further customize. Since this was also the proposal in the Spark community, there's a good chance we can get `ORDER BY` and `DISTRIBUTE BY` into Spark's DDL. It also would not conflict with the meaning of `DISTRIBUTE BY` in Hive.

Reply:
Yes, I intentionally used the definition of `DISTRIBUTE BY`, `SORT BY`, `CLUSTER BY` that both Spark and Hive support in queries and not `CLUSTERED BY` currently present in create table statements. I find it a bit odd that `CLUSTER BY` and `CLUSTERED BY` mean different things. At the same time, [materialized views in Hive](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL#LanguageManualDDL-CreateMaterializedView) do support that syntax.

I went through the discussion on the dev list and have a question. How can I represent a local sort within tasks with the proposed API? I believe it is a common use case and even Iceberg requires that entries for the same partition are next to each other within the same task.

One of my early ideas was to support the following DDLs:

```
...
PARTITIONED BY
DISTRIBUTED BY
SORTED BY
...
```
```
...
PARTITIONED BY
DISTRIBUTED BY
...
```
```
...
PARTITIONED BY
SORTED BY
...
```
```
...
PARTITIONED BY
ORDERED BY
...
```

We can use transforms to add bucketing and we can also support the old `CLUSTERED BY` 
 alone without new clauses for compatibility. Meaning of `SORTED BY` will stay close to the current one. On top, we add `DISTRIBUTED BY` and `ORDERED BY` with their current meaning in Spark.

Then we can map those DDLs into the following write distributions:

```
LocalDistribution
  - sortBy: Seq[Expression]

RangeBasedDistribution
  - distributeBy: Seq[Expression]
  - sortBy: Seq[Expression]

HashBasedDistribution
  - distributeBy: Seq[Expression]
  - sortBy: Seq[Expression]
```

Then `RangeBasedDistribution` can be represented as a global sort depending on `distributeBy` and `sortBy`.

I haven't thought a lot about this, will check the dev discussion again tomorrow.

Reply:
I took a look at the Spark dev discussion one more time and it still looks there is no way to express a local sort without shuffles. After a closer look at Hive, seems they only allow DISTRIBUTE BY and SORT BY together while creating materialized views. In queries, those two can be used independently.

Reply:
There are also certain cases when we write to a table but we don't control the behavior. For example, during eager UPDATEs/DELETEs/MERGEs. Asking for a global sort while writing updated records back might be expensive (apart from the shuffle itself, skew estimation might recompute some parts twice).

Reply:
I think `MERGE` should be handled separately because that will depend on other configuration. You're right that it will also require distribution that is unlikely to match the table's ideal distribution.

Reply:
I don't like using both `SORT BY` and `ORDER BY`. Because there is no obvious difference between them, I think it is confusing. I'd rather use `ORDER BY` for global ordering and maybe add a `LOCAL` keyword to change that: `LOCAL ORDER BY col1, col2`.

If we did that, then we could use `LOCAL ORDER BY` with `DISTRIBUTE BY` or `ORDER BY` on its own. We could also infer that `ORDER BY` is local if `DISTRIBUTE BY` is present. What do you think?